Bug: [v8.0.0-alpha.10] incorrect error message `Parsing error: "parserOptions.programs"`

### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Issue Description

_Bug discussed and confirmed by @JoshuaKGoldberg and @bradzacher._

Config working with `typescript-eslint` v7.9.0 does not work with v8.0.0-alpha.10, raising:
```
/Users/<...>/eslint.config.js
  0:0  error  Parsing error: "parserOptions.programs" has been provided for @typescript-eslint/parser.
The file was not found in any of the provided program instance(s): eslint.config.js
```
although I don't use any home-made program.

Potential culprit: https://github.com/typescript-eslint/typescript-eslint/pull/8834/files

### Reproduction Repository Link

https://github.com/OlivierZal/melcloud-api/blob/e118342a2ce16c49e41e145844bc318e344cfe52/eslint.config.js

### Repro Steps

1) `git clone https://github.com/OlivierZal/melcloud-api`
2) `git co e118342a2ce16c49e41e145844bc318e344cfe52`
3) `eslint .`
4) `npm i typescript-eslint@8.0.0-alpha.10 --force`
5) `eslint .`
6) compare the 2 outputs (see attached screenshots: v7.9.0 vs v8.0.0-alpha)

![v7](https://github.com/typescript-eslint/typescript-eslint/assets/88216225/ec34f899-569f-4b29-9566-488acf22f3a0)
![v8](https://github.com/typescript-eslint/typescript-eslint/assets/88216225/d6924411-191c-41ca-bc1c-9c9fdf5af976)

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `8.0.0-alpha.10` |
| `TypeScript`                           | `5.4.5` |
| `ESLint`                               | `9.2.0` |
| `node`                                 | `22.1.0` |

--- a/packages/typescript-estree/src/create-program/useProvidedPrograms.ts
+++ b/packages/typescript-estree/src/create-program/useProvidedPrograms.ts
@@ -3,9 +3,9 @@ import * as fs from 'fs';
 import * as path from 'path';
 import * as ts from 'typescript';
 
+import type { ParseSettings } from '../parseSettings';
 import type { ASTAndDefiniteProgram } from './shared';
 import { CORE_COMPILER_OPTIONS, getAstFromProgram } from './shared';
-import { ParseSettings } from '../parseSettings';
 
 const log = debug('typescript-eslint:typescript-estree:useProvidedProgram');
 

--- a/packages/typescript-estree/src/create-program/useProvidedPrograms.ts
+++ b/packages/typescript-estree/src/create-program/useProvidedPrograms.ts
@@ -5,45 +5,49 @@ import * as ts from 'typescript';
 
 import type { ASTAndDefiniteProgram } from './shared';
 import { CORE_COMPILER_OPTIONS, getAstFromProgram } from './shared';
+import { ParseSettings } from '../parseSettings';
 
 const log = debug('typescript-eslint:typescript-estree:useProvidedProgram');
 
-export interface ProvidedProgramsSettings {
-  filePath: string;
-  tsconfigRootDir: string;
-}
-
 function useProvidedPrograms(
   programInstances: Iterable<ts.Program>,
-  { filePath, tsconfigRootDir }: ProvidedProgramsSettings,
+  parseSettings: ParseSettings,
 ): ASTAndDefiniteProgram | undefined {
-  log('Retrieving ast for %s from provided program instance(s)', filePath);
+  log(
+    'Retrieving ast for %s from provided program instance(s)',
+    parseSettings.filePath,
+  );
 
   let astAndProgram: ASTAndDefiniteProgram | undefined;
   for (const programInstance of programInstances) {
-    astAndProgram = getAstFromProgram(programInstance, filePath);
+    astAndProgram = getAstFromProgram(programInstance, parseSettings.filePath);
     // Stop at the first applicable program instance
     if (astAndProgram) {
       break;
     }
   }
 
-  if (!astAndProgram) {
-    const relativeFilePath = path.relative(
-      tsconfigRootDir || process.cwd(),
-      filePath,
-    );
-    const errorLines = [
-      '"parserOptions.programs" has been provided for @typescript-eslint/parser.',
-      `The file was not found in any of the provided program instance(s): ${relativeFilePath}`,
-    ];
-
-    throw new Error(errorLines.join('\n'));
+  if (astAndProgram) {
+    astAndProgram.program.getTypeChecker(); // ensure parent pointers are set in source files
+    return astAndProgram;
   }
 
-  astAndProgram.program.getTypeChecker(); // ensure parent pointers are set in source files
+  const relativeFilePath = path.relative(
+    parseSettings.tsconfigRootDir,
+    parseSettings.filePath,
+  );
 
-  return astAndProgram;
+  const [typeSource, typeSources] =
+    parseSettings.projects.size > 0
+      ? ['project', 'project(s)']
+      : ['programs', 'program instance(s)'];
+
+  const errorLines = [
+    `"parserOptions.${typeSource}" has been provided for @typescript-eslint/parser.`,
+    `The file was not found in any of the provided ${typeSources}: ${relativeFilePath}`,
+  ];
+
+  throw new Error(errorLines.join('\n'));
 }
 
 /**

--- a/packages/typescript-estree/tests/lib/semanticInfo.test.ts
+++ b/packages/typescript-estree/tests/lib/semanticInfo.test.ts
@@ -41,6 +41,10 @@ function createOptions(fileName: string): TSESTreeOptions & { cwd?: string } {
 beforeEach(() => clearCaches());
 
 describe('semanticInfo', () => {
+  beforeEach(() => {
+    process.env.TSESTREE_SINGLE_RUN = '';
+  });
+
   // test all AST snapshots
   testFiles.forEach(filename => {
     const code = fs.readFileSync(path.join(FIXTURES_DIR, filename), 'utf8');
@@ -326,6 +330,24 @@ describe('semanticInfo', () => {
       };
       const parseResult = parseAndGenerateServices(code, optionsProjectString);
       expect(parseResult.services.program).toBe(program1);
+    });
+
+    it('file not in single provided project instance in single-run mode should throw', () => {
+      process.env.TSESTREE_SINGLE_RUN = 'true';
+      const filename = 'non-existent-file.ts';
+      const options = createOptions(filename);
+      const optionsWithProjectTrue = {
+        ...options,
+        project: true,
+        programs: undefined,
+      };
+      expect(() =>
+        parseAndGenerateServices('const foo = 5;', optionsWithProjectTrue),
+      ).toThrow(
+        process.env.TYPESCRIPT_ESLINT_PROJECT_SERVICE === 'true'
+          ? `${filename} was not found by the project service. Consider either including it in the tsconfig.json or including it in allowDefaultProject.`
+          : `The file was not found in any of the provided project(s): ${filename}`,
+      );
     });
   }
 

--- a/packages/website/src/hooks/useClipboard.ts
+++ b/packages/website/src/hooks/useClipboard.ts
@@ -9,7 +9,7 @@ export function useClipboard(code: () => string): useClipboardResult {
 
   const copy = useCallback(
     () =>
-      void navigator.clipboard.writeText(code()).then(() => {
+      navigator.clipboard.writeText(code()).then(() => {
         setCopied(true);
       }),
     [setCopied, code],

